File: admin/models/skill_creator.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field
from pi_shared import now_china


class SkillDraft(BaseModel):
    name: str
    description: str
    content: str
    tags: list[str] = Field(default_factory=list)
    # mcp_tools：运行时白名单，精确到工具名（如 wiki_combined_search），由平台按此过滤 MCP 工具
    # Skill 只需描述用到的工具名，不描述工具来自哪个业务 MCP Server（Agent 侧根本看不到 Server 名）
    mcp_tools: list[str] = Field(default_factory=list)
    mcp_tools_reference: str = ""


class SkillCreatorMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str
    created_at: datetime = Field(default_factory=now_china)


class SkillCreatorSession(BaseModel):
    id: str
    user_id: str | None = None
    messages: list[SkillCreatorMessage] = Field(default_factory=list)
    draft: SkillDraft | None = None
    # 发布状态：False 表示草稿进行中，True 表示已发布为 Skill
    published: bool = False
    # 发布后记录对应的 Skill 名称，用于"编辑已保存 Skill"时精确查找会话
    skill_name: str | None = None
    created_at: datetime = Field(default_factory=now_china)
    updated_at: datetime = Field(default_factory=now_china)


class CreateSessionResponse(BaseModel):
    session_id: str
    message: SkillCreatorMessage


class SendMessageRequest(BaseModel):
    content: str


class SendMessageResponse(BaseModel):
    message: SkillCreatorMessage
    draft: SkillDraft | None = None


class PublishSkillRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    content: str | None = None
    tags: list[str] | None = None
    hidden: bool = False


class SkillCreatorUploadResponse(BaseModel):
    """Skill Creator 会话附件上传结果（仅存储，不做后续融合处理）。"""

    filename: str
    relative_path: str = Field(description="相对 skill 目录的路径，如 uploads/doc.pdf")
    stored_path: str = Field(description="落盘绝对路径")
    skill_dir: str = Field(description="目标 skill 目录名（含 _creator/{session_id} 暂存）")
    size: int

File: frontend/src/api/skillCreator.ts
import { mergeTraceHeaders, request } from "./http";
import { downloadBlob } from "../utils/download";
import type { SkillTreeEntry } from "./skills";

export interface SkillDraft {
  name: string;
  description: string;
  content: string;
  tags?: string[];
  /** 运行时白名单，精确到工具名，由 mcp-proxy 按此过滤可用 MCP 工具；不记录工具来自哪个 Server */
  mcp_tools?: string[];
}

/**
 * 与后端 SKILL.md 格式对齐，供右侧预览直接渲染。
 *
 * frontmatter 用 ```yaml 代码块包裹，而不是裸的 `---`：Markdown 渲染器（CommonMark）
 * 把没有空行分隔的连续行当作同一段落，单个换行会被合并成空格，导致 name/description/
 * mcp_tools 等字段挤成一行。代码块内文本按原样保留换行，可以正确逐行展示。
 * 正文 content 已是标准 Markdown（含空行分隔），仍在代码块外正常渲染。
 */
export function buildSkillMarkdown(draft: SkillDraft): string {
  const meta = [`name: ${draft.name}`, `description: ${draft.description}`];
  if ((draft.tags ?? []).length > 0) {
    meta.push("tags:");
    for (const tag of draft.tags ?? []) meta.push(`  - ${tag}`);
  }
  if ((draft.mcp_tools ?? []).length > 0) {
    meta.push("mcp_tools:");
    for (const tool of draft.mcp_tools ?? []) meta.push(`  - ${tool}`);
  }
  return ["```yaml", ...meta, "```", "", draft.content.trim(), ""].join("\n");
}

export interface SkillCreatorMessage {
  role: "user" | "assistant";
  content: string;
  created_at?: string;
}

export interface SkillCreatorSession {
  id: string;
  user_id?: string | null;
  messages: SkillCreatorMessage[];
  draft: SkillDraft | null;
  published: boolean;
  skill_name?: string | null;
  created_at?: string;
  updated_at?: string;
}

export const skillCreatorApi = {
  /**
   * 获取或创建 skill-creator 会话。
   * - skillName 指定时：加载该 Skill 的会话（编辑模式）
   * - forceNew=true：强制新建（新建对话按钮使用）
   * - 默认：复用未发布草稿，无则新建
   */
  openSession: (scope?: "user" | "system", forceNew = false, skillName?: string) => {
    const params = new URLSearchParams();
    if (scope === "user") params.set("scope", "user");
    if (forceNew) params.set("force_new", "true");
    if (skillName?.trim()) params.set("skill_name", skillName.trim());
    const qs = params.toString() ? `?${params.toString()}` : "";
    return request<SkillCreatorSession>(`/config/skills/creator/sessions${qs}`, { method: "POST" });
  },

  getSession: (sessionId: string) =>
    request<SkillCreatorSession>(`/config/skills/creator/sessions/${encodeURIComponent(sessionId)}`),

  resetSession: (sessionId: string) =>
    request<SkillCreatorSession>(`/config/skills/creator/sessions/${encodeURIComponent(sessionId)}/reset`, { method: "POST" }),

  sendMessage: (sessionId: string, content: string, signal?: AbortSignal) =>
    request<{ message: SkillCreatorMessage; draft: SkillDraft | null }>(
      `/config/skills/creator/sessions/${encodeURIComponent(sessionId)}/messages`,
      { method: "POST", body: JSON.stringify({ content }), signal },
    ),

  publish: (sessionId: string, payload: Partial<SkillDraft> & { hidden?: boolean } = {}) =>
    request<{ name: string; description: string; tags?: string[]; hidden?: boolean; user_id?: string | null }>(
      `/config/skills/creator/sessions/${encodeURIComponent(sessionId)}/publish`,
      { method: "POST", body: JSON.stringify(payload) },
    ),

  /** 附件写入对应 skill 目录 uploads/（仅存储，不做融合） */
  uploadFile: async (sessionId: string, file: File) => {
    const form = new FormData();
    form.append("file", file);
    const resp = await fetch(
      `/config/skills/creator/sessions/${encodeURIComponent(sessionId)}/files`,
      {
        method: "POST",
        body: form,
        cache: "no-store",
        headers: mergeTraceHeaders(),
      },
    );
    if (!resp.ok) {
      const err = await resp.json().catch(() => ({}));
      throw new Error((err as { detail?: string }).detail ?? `HTTP ${resp.status}`);
    }
    return resp.json() as Promise<{
      filename: string;
      relative_path: string;
      stored_path: string;
      skill_dir: string;
      size: number;
    }>;
  },

  getTree: (sessionId: string) =>
    request<{ session_id: string; skill_dir: string; entries: SkillTreeEntry[] }>(
      `/config/skills/creator/sessions/${encodeURIComponent(sessionId)}/tree`,
    ),

  fetchFileText: async (sessionId: string, path: string): Promise<string> => {
    const qs = new URLSearchParams({ path, as_text: "true", disposition: "inline" });
    const resp = await fetch(
      `/config/skills/creator/sessions/${encodeURIComponent(sessionId)}/file?${qs}`,
      { cache: "no-store", headers: mergeTraceHeaders() },
    );
    if (!resp.ok) {
      const err = await resp.json().catch(() => ({}));
      throw new Error((err as { detail?: string }).detail ?? `HTTP ${resp.status}`);
    }
    return resp.text();
  },

  fetchFileBlob: async (sessionId: string, path: string): Promise<Blob> => {
    const qs = new URLSearchParams({ path, disposition: "inline" });
    const resp = await fetch(
      `/config/skills/creator/sessions/${encodeURIComponent(sessionId)}/file?${qs}`,
      { cache: "no-store", headers: mergeTraceHeaders() },
    );
    if (!resp.ok) {
      const err = await resp.json().catch(() => ({}));
      throw new Error((err as { detail?: string }).detail ?? `HTTP ${resp.status}`);
    }
    return resp.blob();
  },

  downloadFile: async (sessionId: string, path: string, filename: string): Promise<void> => {
    const qs = new URLSearchParams({ path, disposition: "attachment" });
    const resp = await fetch(
      `/config/skills/creator/sessions/${encodeURIComponent(sessionId)}/file?${qs}`,
      { cache: "no-store", headers: mergeTraceHeaders() },
    );
    if (!resp.ok) {
      const err = await resp.json().catch(() => ({}));
      throw new Error((err as { detail?: string }).detail ?? `HTTP ${resp.status}`);
    }
    const blob = await resp.blob();
    downloadBlob(blob, filename);
  },
};
